iot_config/src/gateway_service.rs

```diff
@@ -1,17 +1,130 @@
-use crate::GrpcResult;
-use helium_proto::services::iot_config::{
-    self, GatewayRegionParamsReqV1, GatewayRegionParamsResV1,
+use crate::{
+    region_map::{self, RegionMap},
+    GrpcResult, Settings,
 };
-use tonic::Request;
+use anyhow::Result;
+use file_store::traits::MsgVerify;
+use helium_crypto::{Keypair, PublicKey, Sign};
+use helium_proto::{
+    services::iot_config::{
+        self, GatewayRegionParamsReqV1, GatewayRegionParamsResV1, LoadRegionReqV1, LoadRegionResV1,
+    },
+    Message, Region,
+};
+use hextree::Cell;
+use node_follower::{
+    follower_service::FollowerService,
+    gateway_resp::{GatewayInfo, GatewayInfoResolver},
+};
+use sqlx::{Pool, Postgres};
+use tonic::{Request, Response, Status};
+
+pub struct GatewayService {
+    admin_pubkey: PublicKey,
+    follower_service: FollowerService,
+    pool: Pool<Postgres>,
+    region_map: RegionMap,
+    signing_key: Keypair,
+}
+
+impl GatewayService {
+    pub async fn new(settings: &Settings) -> Result<Self> {
+        let pool = settings.database.connect(10).await?;
+        let region_map = RegionMap::new(&pool).await?;
+        Ok(Self {
+            admin_pubkey: settings.admin_pubkey()?,
+            follower_service: FollowerService::from_settings(&settings.follower),
+            pool,
+            region_map,
+            signing_key: settings.signing_keypair()?,
+        })
+    }
 
-pub struct GatewayService {}
+    fn verify_admin_signature<R>(&self, request: R) -> Result<R, Status>
+    where
+        R: MsgVerify,
+    {
+        request
+            .verify(&self.admin_pubkey)
+            .map_err(|_| Status::permission_denied("invalid admin signature"))?;
+        Ok(request)
+    }
+}
 
 #[tonic::async_trait]
 impl iot_config::Gateway for GatewayService {
     async fn region_params(
         &self,
-        _request: Request<GatewayRegionParamsReqV1>,
+        request: Request<GatewayRegionParamsReqV1>,
     ) -> GrpcResult<GatewayRegionParamsResV1> {
-        unimplemented!()
+        let request = request.into_inner();
+
+        let pubkey = PublicKey::try_from(request.address.clone())
+            .map_err(|_| Status::invalid_argument("invalid gateway address"))?;
+        request
+            .verify(&pubkey)
+            .map_err(|_| Status::permission_denied("invalid request signature"))?;
+
+        let GatewayInfo { location, gain, .. } = self
+            .follower_service
+            .clone()
+            .resolve_gateway_info(&pubkey.into())
+            .await
+            .map_err(|_| Status::internal("gateway lookup error"))?;
+
+        let location = {
+            let location =
+                location.ok_or_else(|| Status::internal("gateway location undefined"))?;
+            Cell::from_raw(location)
+                .map_err(|_| Status::internal("gateway location is not a valid h3 index"))?
+        };
+
+        let region = self
```

Comment:
I note the request for this API requires a region to be specified.  In the old sibyl gateway APIs that specified region would be used if the gateway was not asserted.  I dont see that same functionality here, Do we need to add that ? If not then we should remove the region field from the request

Reply:
Yeah that needs to be added.

Reply:
so we're saying in the event the gateway doesn't have an asserted location on the chain based on the pubkey lookup then find and return the region params for the region specified in the request?

Reply:
that looks to be the expectation.  In the same  scenario, sibyl would return the specified region, the associated region params and defaulted gain to undefined/none

Reply:
> so we're saying in the event the gateway doesn't have an asserted location on the chain based on the pubkey lookup then find and return the region params for the region specified in the request?

that's correct.. gateway-rs has a default region but no region params.. With this approach it can get region parameters before it has an asserted location on chain

Reply:
implemented in [1634b3c](https://github.com/helium/oracles/pull/311/commits/1634b3c7975c298d09acf50afbfcb1b0150052e9)